Druid/DRABCDConstraintSolver.class.st

```diff
@@ -0,0 +1,188 @@
+
+	"TODO - for unsupported constraints, return true
+			- do we need different ABCD subclasses for upper/lower bounds checks?"
+			
+	| shortestPath |
+	((aDRPiNode constraint isKindOf: DRGreaterOrEqualsConstraint) or:
+		 (aDRPiNode constraint isKindOf: DREqualsConstraint)) ifFalse: [ ^true. ].
+	"TODO podemos hacer por > tambien creo, pidiendo que la distancia sea 1 mas/menos/yoquese"
+		
+	shortestPath := AIShortestPathWithMaxAndMinNodes new.
+	shortestPath nodes: nodes.
+	shortestPath 
+	   edges: edges
+		from: #first
+		to: #second
+		weight: #third.
+
+	shortestPath start: (self piNodesOriginalVariable: aDRPiNode constraint constantValue);
+		isMaxNodePredicate: [ :n | n isPhiFunction ];
+		memoDictionary: memo.
+	shortestPath run.
+
+	"Distance < 0 means the other side is a tautology => this side is unsatisfiable
+	>= 0 means this side isn't tautology => this side is satisfiable"
+	^ (shortestPath distanceTo: aDRPiNode operand isLessThan: 1) = -1.
+	
+]
+
+{ #category : 'building - graph' }
+DRABCDConstraintSolver >> piNodesOriginalVariable: aDRValue [
+
+	| val |
+	
+	val := aDRValue asDRValue.
+	[ val isPiNode ] whileTrue: [ val := val operand ].
+
+	^ val
+]
+
+{ #category : 'building - graph' }
+DRABCDConstraintSolver >> processAdd: aDRValue [
+
+	| ops |
+	ops := aDRValue operands.
+	ops first isNumberValue ifTrue: [ ops := ops reversed ].
+	ops second isNumberValue ifFalse: [ ^ nil ].
+
+	edges add: {
+			ops first.
+			aDRValue.
+			ops second value }
+]
+
+{ #category : 'building - graph' }
+DRABCDConstraintSolver >> processCopy: aDRValue [
+	"Remember a PiNode is a Copy too"
+
+	edges add: {
+			aDRValue operand asDRValue.
+			aDRValue.
+			0 }.
+	
+	nodes add: aDRValue operand asDRValue.
+
+	aDRValue isPiNode ifTrue: [ self processPiNode: aDRValue ]
+]
+
+{ #category : 'building - graph' }
+DRABCDConstraintSolver >> processPhi: aDRPhiFunction [
+
+	aDRPhiFunction operands do: [ :op |
+		nodes add: op.
+		edges add: {
+				op.
+				aDRPhiFunction.
+				0 } ]
+]
+
+{ #category : 'building - graph' }
+DRABCDConstraintSolver >> processPiNode: aDRPiNode [
+
+	| constraint |
+	constraint := aDRPiNode constraint.
+	nodes add: constraint constantValue asDRValue.
+
+	((constraint isKindOf: DRLessOrEqualsConstraint) or:
+		 (constraint isKindOf: DREqualsConstraint)) ifTrue: [
+		^ edges add: {
+				  constraint constantValue asDRValue. "According to the paper, the edge source should be the pi-node associated to the constraint,
+													but I don't think there's a difference here, and this is easier"
+				  aDRPiNode.
+				  0 } ].
+	(constraint isKindOf: DRLessConstraint) ifTrue: [
+		edges add: {
+				constraint constantValue asDRValue. "According to the paper, the edge source should be the pi-node associated to the constraint,
+													but I don't think there's a difference here, and this is easier"
+				aDRPiNode.
+				-1 "I'm assuming integers only, not sure if that's right"} ].
+]
```

Comment:
I'm aware about all these type-checks to manage the constraints. Could we use polymorphism in some way?

I would like something like:
```suggestion
DRABCDConstraintSolver >> processPiNode: aDRPiNode [

	| constraint constant distance |
	constraint := aDRPiNode constraint.
	constant := constraint constantValue asDRValue.
	nodes add: constant.
	
	distance := constraint distanceABCD.
	edges add: {
			constant. "According to the paper, the edge source should be the pi-node associated to the constraint, but I don't think there's a difference here, and this is easier"
			aDRPiNode.
			distance }.
]
```

Reply:
Why here is using `constraint constantValue`, this should not work even between variables?